rpc/jsonrpc/types/types.go

```diff
@@ -47,49 +47,45 @@ func idFromInterface(idInterface interface{}) (jsonrpcid, error) {
 // REQUEST
 
 type RPCRequest struct {
-	JSONRPC string          `json:"jsonrpc"`
-	ID      jsonrpcid       `json:"id,omitempty"`
-	Method  string          `json:"method"`
-	Params  json.RawMessage `json:"params"` // must be map[string]interface{} or []interface{}
+	ID     jsonrpcid
+	Method string
+	Params json.RawMessage
 }
 
-// UnmarshalJSON custom JSON unmarshaling due to jsonrpcid being string or int
+type rpcRequestJSON struct {
+	V  string          `json:"jsonrpc"` // must be "2.0"
+	ID interface{}     `json:"id,omitempty"`
+	M  string          `json:"method"`
+	P  json.RawMessage `json:"params"`
+}
+
+// UnmarshalJSON decodes a request from a JSON-RPC 2.0 request object.
 func (req *RPCRequest) UnmarshalJSON(data []byte) error {
-	unsafeReq := struct {
-		JSONRPC string          `json:"jsonrpc"`
-		ID      interface{}     `json:"id,omitempty"`
-		Method  string          `json:"method"`
-		Params  json.RawMessage `json:"params"` // must be map[string]interface{} or []interface{}
-	}{}
-
-	err := json.Unmarshal(data, &unsafeReq)
-	if err != nil {
+	var wrapper rpcRequestJSON
+	if err := json.Unmarshal(data, &wrapper); err != nil {
 		return err
 	}
-
-	if unsafeReq.ID == nil { // notification
-		return nil
+	// TODO(creachadair): Check version?
```

Comment:
Any reason not to do this now? It's basically free

Reply:
Yes, but it might break existing use. Previously we weren't checking it at all.

Reply:
(that said, I don't know if we should care about existing use that sets this wrong—I'm more concerned with callers that might omit the field entirely)

Reply:
I split the difference and checked that if it's set at all, it must be `"2.0"`, but it's allowed to be omitted.